File: src/main/java/com/evacoffee/beautymod/BeautyMod.java
package com.evacoffee.beautymod;

import com.evacoffee.beautify.network.CustomizationNetworkHandler;
import com.evacoffee.beautify.util.RateLimiter;
import com.evacoffee.beautymod.command.*;
import com.evacoffee.beautymod.components.*;
import com.evacoffee.beautymod.config.ModConfig;
import com.evacoffee.beautymod.config.ModMenuIntegration;
import com.evacoffee.beautymod.data.PlayerDataManager;
import com.evacoffee.beautymod.dating.DateLocationManager;
import com.evacoffee.beautymod.dialogue.DialogueManager;
import com.evacoffee.beautymod.dialogue.DialogueRegistry;
import com.evacoffee.beautymod.entity.ModEntities;
import com.evacoffee.beautymod.entity.NPCRegistry;
import com.evacoffee.beautymod.network.ModPackets;
import com.evacoffee.beautymod.network.SecurePacketHandler;
import com.evacoffee.beautymod.security.AntiHarassmentManager;
import com.evacoffee.beautymod.security.DataEncryption;
import com.evacoffee.beautymod.security.PermissionManager;
import me.shedaniel.autoconfig.AutoConfig;
import me.shedaniel.autoconfig.serializer.JanksonConfigSerializer;
import net.fabricmc.api.ModInitializer;
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback;
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents;
import net.fabricmc.fabric.api.event.player.UseEntityCallback;
import net.fabricmc.loader.api.FabricLoader;
import net.minecraft.entity.passive.VillagerEntity;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import net.minecraft.util.ActionResult;
import net.minecraft.util.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.SecretKey;
import java.nio.file.Path;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

public class BeautyMod implements ModInitializer {
    public static final String MOD_ID = "beautymod";
    public static final String MOD_NAME = "Beauty Mod";
    public static final Logger LOGGER = LoggerFactory.getLogger(MOD_NAME);
    public static final Path CONFIG_DIR = FabricLoader.getInstance().getConfigDir().resolve(MOD_ID);
    
    private static BeautyMod instance;
    private static MinecraftServer serverInstance;
    public static ModConfig CONFIG;

    // Rate limiter for commands and interactions
    private static RateLimiter<UUID> rateLimiter;

    public BeautyMod() {
        instance = this;
    }

    public static BeautyMod getInstance() {
        return instance;
    }

    @Override
    public void onInitialize() {
        LOGGER.info("Initializing {} v{}", MOD_NAME, getModVersion());

        // Create config directory if it doesn't exist
        if (!CONFIG_DIR.toFile().exists()) {
            CONFIG_DIR.toFile().mkdirs();
        }

        // Initialize configuration
        AutoConfig.register(ModConfig.class, JanksonConfigSerializer::new);
        CONFIG = AutoConfig.getConfigHolder(ModConfig.class).getConfig();

        try {
            // Initialize encryption
            SecretKey encryptionKey = DataEncryption.generateKey();
            
            // Initialize core systems
            initializeComponents();
            
            // Initialize player data manager with encryption
            PlayerDataManager.initialize(encryptionKey);
            
            // Register commands
            registerCommands();
            
            // Register event handlers
            registerEventHandlers();
            
            // Initialize networking
            ModPackets.registerCommonPackets();
            
            // Initialize security systems
            AntiHarassmentManager.initialize();
            PermissionManager.initialize();
            
            // Initialize dating system
            DateLocationManager.initialize();
            
            // Initialize NPC registry
            NPCRegistry.initialize();
            
            // Initialize dialogue system
            DialogueRegistry.initialize();
            DialogueManager.initialize();
            
            // Initialize rate limiter (100 requests per second per UUID)
            rateLimiter = new RateLimiter<>(100, 1, TimeUnit.SECONDS);
            
            // Register ModMenu integration
            ModMenuIntegration.register();
            
            LOGGER.info("{} has been successfully initialized!", MOD_NAME);
        } catch (Exception e) {
            LOGGER.error("Failed to initialize " + MOD_NAME, e);
            throw new RuntimeException("Failed to initialize " + MOD_NAME, e);
        }
    }

    private void initializeComponents() {
        // Register all component types
        ModComponents.registerComponentTypes();
    }

    private void registerCommands() {
        CommandRegistrationCallback.EVENT.register((dispatcher, registryAccess, environment) -> {
            // Core commands
            new DatingCommand().register(dispatcher);
            new ProposeCommand().register(dispatcher);
            new DivorceCommand().register(dispatcher);
            new HomeCommand().register(dispatcher);
            new LoveCommand().register(dispatcher);
            new MemoriesCommand().register(dispatcher);
            new ViewMemoriesCommand().register(dispatcher);
            
            // Debug commands (only if enabled in config)
            if (CONFIG.enableDebugCommands) {
                new DebugCommand().register(dispatcher);
                new TalkToNpcCommand().register(dispatcher);
            }
            
            // Security commands (admin only)
            new SecurityCommands().register(dispatcher);
        });
    }
    
    private void registerEventHandlers() {
        // Server lifecycle events
        ServerLifecycleEvents.SERVER_STARTING.register(this::onServerStarting);
        ServerLifecycleEvents.SERVER_STOPPED.register(this::onServerStopped);
        
        // Entity interaction
        UseEntityCallback.EVENT.register((player, world, hand, entity, hitResult) -> {
            if (entity instanceof VillagerEntity) {
                return onVillagerInteract((ServerPlayerEntity) player, (VillagerEntity) entity, hand);
            }
            return ActionResult.PASS;
        });
    }
    
    // Server event handlers
    private void onServerStarting(MinecraftServer server) {
        serverInstance = server;
        LOGGER.info("{} is initializing with server...", MOD_NAME);
        
        // Load data
        try {
            PlayerDataManager.loadAllData();
            PermissionManager.loadPermissions();
            DateLocationManager.loadLocations();
            DialogueManager.loadDialogues();
        } catch (Exception e) {
            LOGGER.error("Failed to load data", e);
        }
    }
    
    private void onServerStopped(MinecraftServer server) {
        LOGGER.info("{} is shutting down...", MOD_NAME);
        
        try {
            // Save data
            PlayerDataManager.saveAllData();
            PermissionManager.savePermissions();
            DateLocationManager.saveLocations();
            DialogueManager.saveDialogues();
        } catch (Exception e) {
            LOGGER.error("Failed to save data", e);
        }
        
        serverInstance = null;
    }
    
    // Interaction handlers
    private ActionResult onVillagerInteract(ServerPlayerEntity player, VillagerEntity villager, net.minecraft.util.Hand hand) {
        // Check if the villager is a romanceable NPC
        if (NPCRegistry.isRomanceNPC(villager)) {
            // Handle romance interaction
            return NPCRegistry.handleInteraction(player, villager, hand);
        }
        return ActionResult.PASS;
    }
    
    // Utility methods
    public static String getModVersion() {
        return FabricLoader.getInstance()
                .getModContainer(MOD_ID)
                .orElseThrow()
                .getMetadata()
                .getVersion()
                .getFriendlyString();
    }
    
    public static MinecraftServer getServer() {
        return serverInstance;
    }
    
    public static RateLimiter<UUID> getRateLimiter() {
        return rateLimiter;
    }
}

File: src/main/kotlin/com/evacoffee/beautymod/BeautyMod.kt
package com.evacoffee.beautymod

import com.evacoffee.beautymod.command.FamilyCommand
import com.evacoffee.beautymod.family.AdoptionManager
import com.evacoffee.beautymod.family.FamilyTree
import com.evacoffee.beautymod.item.AdoptionPapersItem
import net.fabricmc.api.ModInitializer
import net.fabricmc.fabric.api.command.v2.CommandRegistrationCallback
import net.fabricmc.fabric.api.event.lifecycle.v1.ServerLifecycleEvents
import net.fabricmc.fabric.api.itemgroup.v1.FabricItemGroup
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemStack
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.server.MinecraftServer
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.util.*

object BeautyMod : ModInitializer {
    // Mod metadata
    const val MOD_ID = "beautymod"
    const val MOD_NAME = "BeautyMod"
    val LOGGER: Logger = LoggerFactory.getLogger(MOD_NAME)
    
    // Data files
    private const val DATA_DIR = "beautymod"
    private const val ADOPTION_DATA = "adoptions.dat"
    
    // Networking
    val ADOPTION_REQUEST_PACKET = Identifier(MOD_ID, "adoption_request")
    val ADOPTION_RESPONSE_PACKET = Identifier(MOD_ID, "adoption_response")
    
    // Managers
    lateinit var adoptionManager: AdoptionManager
        private set
    lateinit var familyTree: FamilyTree
        private set
    
    // Item Group
    val ITEM_GROUP: ItemGroup = FabricItemGroup.builder()
        .icon { ItemStack(Items.ADOPTION_PAPERS) }
        .displayName(Text.translatable("itemGroup.beautymod.main"))
        .build()
    
    // Items
    object Items {
        val ADOPTION_PAPERS = AdoptionPapersItem
    }
    
    // Server instance
    private var server: MinecraftServer? = null
    
    override fun onInitialize() {
        LOGGER.info("Initializing $MOD_NAME")
        
        // Register item group
        Registry.register(
            Registries.ITEM_GROUP,
            Identifier(MOD_ID, "main"),
            ITEM_GROUP
        )
        
        // Register items
        Items.ADOPTION_PAPERS // This registers the item through its init block
        
        // Register commands
        CommandRegistrationCallback.EVENT.register { dispatcher, _, _ ->
            FamilyCommand.register(dispatcher)
        }
        
        // Server lifecycle events
        ServerLifecycleEvents.SERVER_STARTING.register { server ->
            this.server = server
            loadData(server)
        }
        
        ServerLifecycleEvents.SERVER_STOPPING.register { server ->
            saveData(server)
            this.server = null
        }
        
        // Register network receivers
        registerNetworkHandlers()
        
        LOGGER.info("$MOD_NAME has been initialized!")
    }
    
    private fun loadData(server: MinecraftServer) {
        val saveDir = File(server.runDirectory, DATA_DIR)
        if (!saveDir.exists()) {
            saveDir.mkdirs()
        }
        
        val adoptionFile = File(saveDir, ADOPTION_DATA)
        adoptionManager = AdoptionManager()
        familyTree = FamilyTree()
        
        if (adoptionFile.exists()) {
            try {
                val nbt = net.minecraft.nbt.NbtIo.readCompressed(adoptionFile)
                adoptionManager.readNbt(nbt)
                LOGGER.info("Loaded adoption data")
            } catch (e: Exception) {
                LOGGER.error("Failed to load adoption data", e)
            }
        }
    }
    
    private fun saveData(server: MinecraftServer) {
        val saveDir = File(server.runDirectory, DATA_DIR)
        if (!saveDir.exists()) {
            saveDir.mkdirs()
        }
        
        try {
            val nbt = net.minecraft.nbt.NbtCompound()
            adoptionManager.writeNbt(nbt)
            net.minecraft.nbt.NbtIo.writeCompressed(nbt, File(saveDir, ADOPTION_DATA))
            LOGGER.info("Saved adoption data")
        } catch (e: Exception) {
            LOGGER.error("Failed to save adoption data", e)
        }
    }
    
    private fun registerNetworkHandlers() {
        // Handle adoption request responses from clients
        ServerPlayNetworking.registerGlobalReceiver(ADOPTION_RESPONSE_PACKET) { server, player, _, buf, _ ->
            val accepted = buf.readBoolean()
            server.execute {
                if (accepted) {
                    adoptionManager.acceptAdoption(player)
                } else {
                    adoptionManager.denyAdoption(player)
                }
            }
        }
    }
    
    // Helper function to get the server instance
    fun getServer(): MinecraftServer? = server
}